test/UpgradeRegressionManager.invariant.t.sol

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.30;
+
+// External Imports
+import {Test} from "forge-std/Test.sol";
+
+// Internal Imports
+import {UpgradeRegressionManager} from "src/contracts/UpgradeRegressionManager.sol";
+import {MockTimelockTarget} from "test/mocks/MockTimelockTarget.sol";
+import {SimpleContract} from "test/helpers/SimpleContract.sol";
+import {UpgradeRegressionManagerHandler} from "test/handlers/UpgradeRegressionManager.handler.sol";
+import {ITimelockTarget} from "src/interfaces/ITimelockTarget.sol";
+
+contract UpgradeRegressionManagerInvariantTest is Test {
+  UpgradeRegressionManager public urm;
+  MockTimelockTarget public timelockTarget;
+  SimpleContract public target;
+  UpgradeRegressionManagerHandler public handler;
+
+  address public admin = makeAddr("admin");
+  address public guardian = makeAddr("guardian");
+
+  function setUp() public {
+    timelockTarget = new MockTimelockTarget();
+    timelockTarget.setDelay(5 seconds);
+
+    target = new SimpleContract();
+
+    urm = new UpgradeRegressionManager(ITimelockTarget(timelockTarget), admin, guardian, 5 seconds);
+
+    handler = new UpgradeRegressionManagerHandler(urm, admin, guardian, target);
+
+    // target the handler for invariant testing
+    targetContract(address(handler));
+  }
+
+  /*///////////////////////////////////////////////////////////////
+                        Rollback State Invariant
+  //////////////////////////////////////////////////////////////*/
+
+  /// @notice A rollback cannot exist in both `rollbackQueueExpiresAt` and `rollbackExecutableAt` simultaneously
+  /// @dev Rationale: Ensures rollback is in exactly one state at a time
+  function invariant_rollbackStateMutualExclusion() public view {
+    uint256 rollbackId = handler.getRollbackId();
+
+    bool inQueueingState = urm.rollbackQueueExpiresAt(rollbackId) != 0;
+    bool inExecutionState = urm.rollbackExecutableAt(rollbackId) != 0;
+
+    assertFalse(inQueueingState && inExecutionState, "Rollback cannot exist in both queueing and execution state");
+  }
+
+  /*///////////////////////////////////////////////////////////////
+                        Timing Invariant
+  //////////////////////////////////////////////////////////////*/
+
+  /// @notice `rollbackQueueWindow` must always be greater than 0
+  /// @dev Ensures valid queue window for rollbacks
+  function invariant_rollbackQueueWindowPositive() public view {
```

Comment:
Seeing it implemented, this feels like overkill. Input validation is adequately covered by a unit test, I think.